### PR TITLE
Fix lora creation

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?
1. Makes the injecting of LoRA layers faster. The previous way of checking `target_module_found` was the bottleneck. This PR fixes that.

This help the PR https://github.com/huggingface/diffusers/pull/5151 when loading large SD-XL model. The time to load the lora ckpts went down from ~120 seconds to ~2 seconds.

Screenshot 2023-10-04 at 6 38 05 PM

### CODE REVIEW

File: src/peft/tuners/lora/model.py [hunk 1]
```diff
@@ -493,19 +495,19 @@ def add_weighted_adapter(
...
                 )
-            if target_modules_type == list:
-                new_target_modules |= set(self.peft_config[adapter].target_modules)
+            if target_modules_type == set:
+                new_target_modules |= self.peft_config[adapter].target_modules
             else:
                 new_target_modules += f"({self.peft_config[adapter].target_modules})|"
 
-        new_target_modules = list(new_target_modules) if target_modules_type == list else new_target_modules[:-1]
+        new_target_modules = list(new_target_modules) if target_modules_type == set else new_target_modules[:-1]
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/model.py:510]: I think this whole block could be simplified and made more robust like so:

```python
        target_module_types = [type(self.peft_config[adapter].target_modules) for adapter in adapters]
        if not target_module_types:
            raise ValueError(f"Found no adapter matching the names in {adapters}")
        if len(set(target_module_types)) > 1:
            raise ValueError(
                "all adapter configs should follow the same target modules type. "
                "Combining adapters with `target_modules` type being a mix of list/set and string is not supported."
            )

        if target_module_types[0] == str:
            new_target_modules += "|".join(f"({self.peft_config[adapter].target_modules})" for adapter in adapters)
        elif target_module_types[0] == set:
            new_target_modules = reduce(operator.or_, (self.peft_config[adapter].target_modules for adapter in adapters))
        else:
            raise TypeError(f"Invalid type {target_module_types[0]} found in target_modules")
```

This requires `from functools import reduce` and `import operator`.

File: src/peft/tuners/tuners_utils.py [hunk 2]
```diff
@@ -359,8 +359,8 @@ def check_target_module_exists(config, key: str) -> bool | re.Match[str] | None:
     if isinstance(config.target_modules, str):
         target_module_found = re.fullmatch(config.target_modules, key)
     else:
-        target_module_found = any(re.match(f".*\.{target_key}$", key) for target_key in config.target_modules) or any(
-            target_key == key for target_key in config.target_modules
+        target_module_found = key in config.target_modules or any(
+            key.endswith(f".{target_key}") for target_key in config.target_modules
```
Comment (BenjaminBossan) [inline src/peft/tuners/tuners_utils.py:363]: Could you please explain why this change was made? Is it related to the other changes in this PR or just to simplify?

**Review by BenjaminBossan — COMMENTED** [review]
Great discovery of the performance improvement and nice changes. I have some comments, please check them out.

File: src/peft/tuners/lora/model.py (hunk 1 above)
Comment (pacman100) [reply src/peft/tuners/lora/model.py]: Thank you so much for this. Incorporated this in the latest commit.

**Review by BenjaminBossan — APPROVED** [review]
LGTM, thanks for implementing this performance increase.
